man/hastec.1
.TH HASTEC 1

.SH NAME
hastec \- compile Haskell programs into JavaScript

.SH SYNOPSIS
.B hastec
.RI [ option | filename ]...

.SH DESCRIPTION
This manual page documents briefly the
.B hastec
command.
In addition to the options listed here,
.B hastec
also accepts all the same command line options as the GHC compiler,
with the exception of
.B -prof
which is silently ignored.

.BR
By default,
.B hastec
behaves as though it was called with GHC's
.B --make
option, compiling and linking a complete JavaScript program which,
when included into a web page, sets the program's
.B main
function to execute when the page has finished loading.



.SH OPTIONS
Options are processed from left to right. If two conflicting options are given,
the latter will take precedence. This means that you can use options such as
.B --opt-all
which imply multiple other options, and selectively disable any
implied options that you don't want. For instance, the command line
.B hastec --opt-all --opt-tail-loop-transform=off
will enable all safe optimizations except for the tail loop transformation.


.SS Linking

.TP
.BR \-\-dont\-link
Don't link generated .jsmod files into a .js blob.

.TP
.BR \-\-link\-jslib [=\fIFILE\fR]
Create a jslib file instead of an executable. If no FILE is given, the package
key of the library currently being compiled is used as the library file name.

.TP
.BR \-\-with\-js =\fIFILES\fR
Link the given comma-separated list of JavaScript files into the final .js file.


.SS Program startup

.TP
.BR \-\-onexec
Launch application immediately when the JavaScript file is loaded.
Shorthand for
.B --start=onexec.

.TP
.BR \-\-onload
Launch application on window.onload. Shorthand for
.B --start=onload.
This is the default behavior.

.TP
.BR \-\-start =\fICODE\fR
Specify custom start code. '$HASTE_MAIN' will be replaced with the
application's main function. For instance,
.B --start='$("foo").onclick($HASTE_MAIN);'
will use jQuery to launch the
application whenever the element with the id "foo" is clicked.


.SS Output control

.TP
.BR \-o\fIFILE\fR ", " \-\-out =\fIFILE\fR
Write JavaScript output to FILE.

.TP
.BR \-\-outdir =\fIDIR\fR
Write intermediate files to DIR.

.TP
.BR \-\-output\-html
Write the JavaScript output to an HTML file with a simple HTML skeleton.

.TP
.BR \-\-own\-namespace
Wrap the whole program in a closure to avoid polluting the global namespace.
Incurs a performance hit, and makes minification slightly less effective.


.SS Code generation

.TP
.BR \-\-full\-unicode
Enable full generalCategory Unicode support. May bloat output by upwards of
150 KB.

.TP
.BR \-\-no\-use\-strict
Do not emit '"use strict";' declaration. Does not affect minifier behavior,
but
.B does
affect any external JavaScript included using
.B --with-js
\.

.TP
.BR \-\-overwrite\-scrutinees
Overwrite scrutinees when evaluated rather than allocating a new local for
the evaluated value. This is largely experimental.

.TP
.BR \-\-output\-jsflow
Output code for use with the JSFlow interpreter. Note that this may leave
your code crippled, since JSFlow doesn't support all of Haste's needs.


.SS Debugging

.TP
.BR \-\-annotate\-externals
Annotate all JavaScript-native symbols and inline JavaScript in
generated code with /* EXTERNAL */.

.TP
.BR \-\-annotate\-symbols
Annotate all non-external, non-local symbols with their qualified Haskell names.

.TP
.BR \-\-debug
Output annotated, pretty-printed JavaScript code.
Equivalent to
.B --annotate-externals --annotate-symbols --pretty-print
\.

.TP
.BR \-\-ddisable\-js\-opts
Disable any and all optimizations over the resulting JavaScript code.
This notably includes tail call optimization.

.TP
.BR \-\-dtrace\-primops
Trace primops.
Not really useful unless Haste was booted with primop tracing enabled.

.TP
.BR \-\-preserve\-names
Preserve Haskell names in JavaScript code as far as possible.
Highly experimental and may break your code.

.TP
.BR \-\-pretty\-print
Print JavaScript output using whitespace and indentation.


.SS Optimization

.TP
.BR \-\-opt\-all
Enable all safe optimizations except minification. Individual optimizations
may be turned off using their individual flags.

.TP
.BR \-\-opt\-detrampoline\-threshold =\fIN\fR
Remove trampolining and tail calls for provably finite tail call chains shorter
than \fIN\fR calls. Set to \fI0\fR to disable. A value of \fIN=3\fR is
implied by
.B --opt-all.

.TP
.BR \-\-opt\-flow-analysis [=\fIon\fR/\fIoff\fR]
Enable whole program flow analysis. Highly experimental and possibly slow
and/or incorrect. Don't use for now.

.TP
.BR \-\-opt\-inline-ffi\-primitives [=\fIon\fR/\fIoff\fR]
Inline FFI call primitives where possible.

.TP
.BR \-\-opt\-minify [=\fIon\fR/\fIoff\fR]
Minify JavaScript output using Google Closure compiler.

.TP
.BR \-\-opt\-minify\-flag =\fIFLAG\fR
Pass a flag to Closure. To minify programs in strict mode, use
.B --opt-minify-flag='--language_in=ECMASCRIPT5_STRICT'.

.TP
.BR \-\-opt\-tail-chain-bound =\fIN\fR
Bound tailcall chains at \fIN\fR stack frames.
By default, tail call chains are not allowed to grow the call stack at all.
This is very inefficient, however, so ideally one wants to strike a balance
between amortizing the cost of the trampolining machinery and the size of the
call stack. A value between 10 and 100 is usually a good choice.

.TP
.BR \-\-opt\-tail\-loop\-transform [=\fIon\fR/\fIoff\fR]
Optimize tail recursive functions into loops when possible.
Enabled by default.

.TP
.BR \-\-opt\-unsafe
Enable all optimizations, safe and unsafe. Equivalent to
.B --opt-all --opt-unsafe-ints.

.TP
.BR \-\-opt\-unsafe\-ints
Enable unsafe Int arithmetic. Implies
.B --opt-unsafe-mult --opt-vague-ints.

.TP
.BR \-\-opt\-unsafe\-mult [=\fIon\fR/\fIoff\fR]
Use JavaScript's built-in multiplication operator for fixed precision integer
multiplication. This may speed up Int multiplication by a factor of at least
four, but may give incorrect results when the product falls outside the
interval [-2^52, 2^52]. In browsers which support Math.imul, this
optimization will likely be slower than the default.

.TP
.BR \-\-opt\-vague\-ints [=\fIon\fR/\fIoff\fR]
Int math has 53 bits of precision, but gives incorrect results rather than
properly wrapping around when those 53 bits are exceeded. Bitwise operations
still only work on the lowest 32 bits.

.TP
.BR \-\-opt\-whole\-program [=\fIon\fR/\fIoff\fR]
Perform optimizations over the whole program during linking. May
significantly increase link time.


.SS Misc. options

.TP
.BR \-? ", " \-\-help
Display help message.

.TP
.BR \-v ", " \-\-verbose
Display even the most obnoxious warnings and messages.


debian/changelog
haste-compiler (0.5.3) UNRELEASED; urgency=medium

  * New APIs. Also bug fixes.

 -- Anton Ekblad <anton@ekblad.cc>  Mon, 05 Oct 2015 15:49:57 +0200


src/hastec.hs
{-# LANGUAGE CPP #-}
-- | Haste's main compiler driver.
module Main where
import Language.Haskell.GHC.Simple
#if __GLASGOW_HASKELL__ >= 710
import Language.Haskell.GHC.Simple.PrimIface
import Packages
import PackageConfig
#endif
import GHC
import Outputable (showPpr)
import Platform

import System.Environment (getArgs, lookupEnv, setEnv)
import System.Exit
import Data.List
import System.IO.Unsafe
import System.IO
import System.FilePath
import System.Directory (copyFile)
import Control.Monad
import qualified Control.Shell as Sh
import qualified Data.ByteString.UTF8 as BS

import Haste.Opts
import Haste.Args
import Haste.Config
import Haste.Environment
import Haste.Version
import Haste.Module
import Haste.CodeGen
import Haste.Linker
import Haste.JSLib

logStr :: Config -> String -> IO ()
logStr cfg = when (verbose cfg) . hPutStrLn stderr

main :: IO ()
main = do
    initUserPkgDB
    as <- getArgs
    lookupEnv "HASTE_PACKAGE_PATH" >>= setEnv "GHC_PACKAGE_PATH" . maybe "" id
    booting <- maybe False (const True) `fmap` lookupEnv "HASTE_BOOTING"
    let args = "-O2":concat [pkgs booting, as, ["-D__HASTE__="++show intVersion]]
        prof = "-prof" `elem` args
    case parseHasteFlags booting args as of
      Left act             -> act
      Right (fs, mkConfig) -> do
        let ghcconfig = mkGhcCfg fs args
        (dfs, _) <- getDynFlagsForConfig ghcconfig
        extralibdirs <- getExtraLibDirs dfs
        let cfg = mkLinkerCfg dfs extralibdirs
                . setShowOutputable dfs
                $ mkConfig def
        res <- compileFold ghcconfig (compJS cfg) ([], []) []
        case res of
          Failure _ _         -> do
            exitFailure
          Success (targets, mods) _ _ -> do
            when (performLink cfg) $ do
              if linkJSLib cfg
                then buildJSLib prof cfg (fst $ head targets) mods
                else mapM_ (uncurry $ linkAndMinify cfg) targets
  where
#if __GLASGOW_HASKELL__ >= 710
    getExtraLibDirs = fmap (concatMap libraryDirs) . readPackageConfigs
#else
    getExtraLibDirs = const (return [])
#endif

    dotToSlash '.' = '/'
    dotToSlash c   = c

    pkgs booting =
      ["-no-global-package-db",
       "-no-user-package-db",
       "-package-db=" ++ pkgSysDir] ++
      if booting && portableHaste
        then []
        else ["-package-db=" ++ pkgUserDir]

    buildJSLib profiling cfg pkgkey mods = do
      let modpath     = targetLibPath cfg ++ "/" ++ pkgkey
          mods'       = [ modpath ++ "/" ++ map dotToSlash mn ++ ".jsmod"
                        | mn <- mods]
          libfile     = maybe (pkgkey <.> "jsmod") id (linkJSLibFile cfg)
          profLibfile = reverse (drop 6 (reverse libfile)) ++ "_p.jslib"

      createJSLib libfile (BS.fromString pkgkey) mods'
      when profiling . void $ do
        copyFile libfile (profLibfile)

    compJS cfg (targets, mods) m = do
      compJSMod cfg m
      let infile = maybe (modInterfaceFile m) id (modSourceFile m)
          modpair = (modPackageKey m, infile)
      if modIsTarget m
        then return $ (modpair : targets, modName m : mods)
        else return (targets, modName m : mods)

    mkGhcCfg fs args = disableCodeGen $ defaultConfig {
        cfgGhcFlags = fs,
        cfgGhcLibDir = Just hasteGhcLibDir,
        cfgUseTargetsFromFlags = True,
        cfgUseGhcErrorLogger = True,
        cfgUpdateDynFlags = \dfs -> dfs {
            ghcMode = if "-c" `elem` args
                        then OneShot
                        else CompManager,
            settings = (settings dfs) {
                sTargetPlatform = (sTargetPlatform $ settings dfs) {
                    platformArch             = ArchX86,
                    platformWordSize         = 4
                  },
                sPlatformConstants = (sPlatformConstants $ settings dfs) {
                    pc_WORD_SIZE       = 4,
                    pc_CINT_SIZE       = 4,
                    pc_CLONG_SIZE      = 4,
                    pc_CLONG_LONG_SIZE = 8,
                    pc_DOUBLE_SIZE     = 8,
                    pc_WORDS_BIGENDIAN = False
                  }
             }
          }
#if __GLASGOW_HASKELL__ >= 710
        , cfgCustomPrimIface = Just (primOpInfo, primOpStrictness)
#endif
      }
    mkLinkerCfg dfs extralibdirs cfg = cfg {
        mainMod = Just (pkgKeyString $ modulePkgKey (mainModIs dfs),
                        moduleNameString $ moduleName (mainModIs dfs)),
        libPaths = extralibdirs ++ libPaths cfg
      }
    setShowOutputable dfs cfg = cfg {showOutputable = showPpr dfs}

#if __GLASGOW_HASKELL__ >= 710
-- | Primop info for custom GHC.Prim interface.
primOpInfo :: PrimOp -> PrimOpInfo
#include "primop-info-710.hs"

-- | Strictness info for custom GHC.Prim interface.
primOpStrictness :: PrimOp -> Arity -> StrictSig
#include "primop-stricts-710.hs"
#endif

-- | Compile an STG module into a JS module and write it to its appropriate
--   location according to the given config.
compJSMod :: Config -> StgModule -> IO ()
compJSMod cfg stg = do
    logStr cfg $ "Compiling " ++ myName ++ " into " ++ targetpath
    writeModule targetpath (generate cfg stg) boot
  where
    boot = modSourceIsHsBoot stg
    myName = modName stg ++ if boot then " [boot]" else ""
    targetpath = targetLibPath cfg

-- | Link a program starting from the 'mainMod' symbol of the given 'Config'.
--   Minify the result if indicated by the config.
linkAndMinify :: Config -> String -> FilePath -> IO ()
linkAndMinify cfg pkgkey infile = do
    logStr cfg $ "Linking target " ++ outfile
    link cfg (BS.fromString pkgkey) infile
    case useGoogleClosure cfg of
      Just clopath -> closurize cfg clopath outfile
      _            -> return ()
    when (outputHTML cfg) $ do
      res <- Sh.shell $ Sh.withCustomTempFile "." $ \tmp h -> do
        prog <- Sh.file outfile
        Sh.hPutStrLn h (htmlSkeleton outfile prog)
        Sh.liftIO $ hClose h
        Sh.mv tmp outfile
      case res of
        Right () -> return ()
        Left err -> error $ "Couldn't output HTML file: " ++ err
  where
    outfile = outFile cfg cfg infile

-- | Produce an HTML skeleton with an embedded JS program.
htmlSkeleton :: FilePath -> String -> String
htmlSkeleton filename prog = concat [
  "<!DOCTYPE HTML>",
  "<html><head>",
  "<title>", filename , "</title>",
  "<meta charset=\"UTF-8\">",
  "<script type=\"text/javascript\">", prog, "</script>",
  "</head><body></body></html>"]

-- | Run Google Closure on a file.
closurize :: Config -> FilePath -> FilePath -> IO ()
closurize cfg cloPath f = do
  let arguments = useGoogleClosureFlags cfg
  logStr cfg $ "Minifying " ++ f ++ "..."
  let cloFile = f `Sh.addExtension` ".clo"
  res <- Sh.shell $ do
    str <- Sh.run "java"
      (["-jar", cloPath,
        "--compilation_level", "ADVANCED_OPTIMIZATIONS",
        "--jscomp_off", "globalThis", f]
       ++ arguments) ""
    Sh.file cloFile str :: Sh.Shell ()
    Sh.mv cloFile f
  case res of
    Left e  -> fail $ "Couldn't execute Google Closure compiler: " ++ e
    Right _ -> return ()

-- | Call vanilla GHC; used for C files and the like.
callVanillaGHC :: [String] -> IO ()
callVanillaGHC args = do
    booting <- maybe False (const True) `fmap` lookupEnv "HASTE_BOOTING"
    _ <- Sh.shell $ Sh.run_ ghcBinary (pkgargs booting ++ ghcArgs) ""
    return ()
  where
    Right (_, ghcArgs, _) = parseArgs hasteOpts "" args
    pkgargs booting =
      ["-no-global-package-db",
       "-no-user-package-db",
       "-package-db=" ++ pkgSysDir] ++
      if booting && portableHaste
        then []
        else ["-package-db=" ++ pkgUserDir]

-- | Initialize the Haste package database, unless it already exists.
initUserPkgDB :: IO ()
initUserPkgDB = do
  _ <- Sh.shell $ do
    pkgDirExists <- Sh.isDirectory pkgUserDir
    when (not pkgDirExists) $ do
      Sh.mkdir True pkgUserLibDir
      Sh.runInteractive hastePkgBinary ["init", pkgUserDir]
  return ()

type Message = String
data Compiler = Haste | GHC | InstallJSExe | BuildRunner deriving (Show, Eq)
data RunMode = Run !Compiler | DontRun !Message deriving (Show, Eq)

rebootMsg :: Message
rebootMsg = "Haste needs to be rebooted; please run haste-boot"

-- | How should we run the compiler for this command line?
runMode :: Bool -> [String] -> RunMode
runMode booting as
  | "--help" `elem` as                    = Run Haste
  | "--install-executable" `elem` as      = Run InstallJSExe
  | "--build-runner" `elem` as            = Run BuildRunner
  | "--info" `elem` as                    = DontRun ghcInfo
  | "--print-libdir" `elem` as            = DontRun hasteGhcLibDir
  | "--print-global-package-db" `elem` as = DontRun pkgSysDir
  | "--version" `elem` as                 = DontRun $ showVersion hasteVersion
  | "--numeric-ghc-version" `elem` as     = DontRun $ showVersion ghcVersion
  | "--supported-extensions" `elem` as    = DontRun exts
  | "--supported-languages" `elem` as     = DontRun exts
  | hasteNeedsReboot && not booting       = DontRun rebootMsg
  | otherwise                             = Run (chooseCompilerFor as)
  where
    exts = unlines supportedLanguagesAndExtensions
    ghcInfo = unsafePerformIO $ do
      dfs <- runGhc (Just hasteGhcLibDir) $ getSessionDynFlags
      return $ formatInfo $ compilerInfo dfs
    formatInfo = ('[' :) . tail . unlines . (++ ["]"]) . map ((',' :) . show)

-- | Parse Haste and static GHC flags, returning either an action to be taken
--   before promptly exiting, or a Haste config and a list of flags for GHC.
parseHasteFlags :: Bool -> [String] -> [String]
                -> Either (IO ()) ([String], Config->Config)
parseHasteFlags booting args rawargs = do
  case runMode booting args of
    DontRun msg -> Left $ putStrLn msg
    Run GHC     -> Left $ callVanillaGHC args
    Run Haste   -> do
      case parseArgs hasteOpts helpHeader args of
        Left msg              -> Left $ putStrLn msg
        Right (cfg, rest, []) -> Right (filter (/= "-prof") rest, cfg)
        Right (_, _,      es) -> Left (mapM_ putStr es >> exitFailure)
    Run InstallJSExe -> do
      Left $ installJSExe (jsexeIn args) (jsexeOut args)
    Run BuildRunner -> do
      Left $ callVanillaGHC (rawargs ++ ["-package-id", "Cabal-1.23.0.0-f701f4ea98ec7bed5883c4df743045e6"])
  where
     jsexeIn ("--install-executable":exe:_) = exe
     jsexeIn (_:xs)                         = jsexeIn xs
     jsexeIn _                              = error "No executable to install!"

     jsexeOut ("-o":out:_) = out
     jsexeOut (_:xs)       = jsexeOut xs
     jsexeOut _            = error "No executable install target!"

-- | Install executable as an HTML file.
installJSExe :: FilePath -> FilePath -> IO ()
installJSExe inf outf = do
  js <- readFile inf
  if (take 2 js == "<!")
    then writeFile (replaceExtension outf "html") js
    else writeFile (replaceExtension outf "html") (htmlSkeleton outf js)

-- | Use Haste or GHC for this command line?
chooseCompilerFor :: [String] -> Compiler
chooseCompilerFor args
  | all hasteOK args  = Haste
  | otherwise         = GHC
  where
    hasteOK f = not $ any (`isSuffixOf` f) [".c",".cmm",".cc",".dyn_o"]
